src/services/user-centrics-service.ts

```diff
@@ -0,0 +1,51 @@
+import { UCUICMPEvent, UCUICMPEventType } from '../types'
+
+export type ServiceState = {
+    initialized: boolean
+    isOpen: boolean
+}
+
+const isUCUICMPEvent = (event: Event): event is UCUICMPEvent => event.type === 'UC_UI_CMP_EVENT'
+
+const state: ServiceState = {
+    initialized: false,
+    isOpen: false,
+}
+
+const subscriptions: ((x: ServiceState) => void)[] = []
+const update = () => subscriptions.forEach((s) => s(state))
+const subscribe = (x: (x: ServiceState) => void) => subscriptions.push(x)
+
+const setIsOpen = (x: boolean) => {
+    state.isOpen = x
+    update()
+}
+const setIsInitialized = (x: boolean) => {
+    state.initialized = x
+    update()
+}
+
+const onEvent = (event: Event) => {
+    if (isUCUICMPEvent(event)) {
+        switch (event.detail.type) {
+            case UCUICMPEventType.CMP_SHOWN:
+                setIsOpen(true)
+                break
+            case UCUICMPEventType.ACCEPT_ALL:
+            case UCUICMPEventType.DENY_ALL:
+            case UCUICMPEventType.SAVE:
+                setIsOpen(false)
+                break
+        }
+    }
+}
+
+if (typeof window !== 'undefined') {
+    window.addEventListener('UC_UI_CMP_EVENT', onEvent)
+    window.addEventListener('UC_UI_INITIALIZED', () => setIsInitialized(true))
+}
+
+export const UserCentericsService = {
+    state,
+    subscribe,
+}
```

Comment:
We could encapsulate the `state` and `subscriptions` states by making this a factory. This would be a bit cleaner in my opinion. Something like:

```ts
const usercentrics = createUsercentricsService()
```

where all things currently on module-level are moved into the function body.

What do you think?

Reply:
It is currently a singleton yes, but does it have to be? I don't really have a strong opinion on this, so just a thought.

Reply:
Let's keep it as a singleton for now.